### ISSUE TITLE
Inconsistencies in testing

### ISSUE THREAD

**Issue by NiMlr** [mdp-toolkit/mdp-toolkit#58]
At least in some [test session](https://travis-ci.org/mdp-toolkit/mdp-toolkit/jobs/624711958?utm_medium=notification&utm_source=github_status)s, not all dependencies are installed. I am not sure about the consequences of this.

```
============================= test session starts ==============================
platform linux -- Python 3.6.7, pytest-4.0.0, py-1.7.0, pluggy-0.8.0
          python: 3.6.7.final.0
             mdp: 3.5, MDP-3.5-LKG-140-g9251ea7
 parallel python: NOT AVAILABLE: No module named 'pp'
          shogun: NOT AVAILABLE: No module named 'shogun'
          libsvm: NOT AVAILABLE: No module named 'svm'
          joblib: 0.14.1
         sklearn: 0.21.2
            numx: scipy 1.3.3
          symeig: scipy.linalg.eigh
```

This [dates back at least 2 years](https://travis-ci.org/mdp-toolkit/mdp-toolkit/builds/275806944?utm_medium=notification&utm_source=github_status).

This phenomenon appears depending on version combinations as well. See that [Python: 3.7-dev, VNUMPY=1.16.1, VSCI=0.20.0](https://travis-ci.org/mdp-toolkit/mdp-toolkit/jobs/624711967?utm_medium=notification&utm_source=github_status) properly installs sklearn, whereas [Python: 3.6, VNUMPY=1.12.1, VSCI=0.20.0](https://travis-ci.org/mdp-toolkit/mdp-toolkit/jobs/624711954?utm_medium=notification&utm_source=github_status) does not.

Additionally, in [Python: 3.7-dev, VNUMPY=1.16.1, VSCI=0.20.0](https://travis-ci.org/mdp-toolkit/mdp-toolkit/jobs/624711967?utm_medium=notification&utm_source=github_status) the backend is provided by scipy and in [Python: 3.6, VNUMPY=1.12.1, VSCI=0.20.0](https://travis-ci.org/mdp-toolkit/mdp-toolkit/jobs/624711954?utm_medium=notification&utm_source=github_status) the numx backend is provided by numpy.

Possibly, there should be a careful overhaul of the testing setup. I am happy to work on this, but it could be helpful to have some assistance by a more senior MDP developer. Ideally, one that has been involved in the testing previously.

When tackling this issue it may be interesting to look at a guideline on the minimum passing probability for random tests. An issue with this [popped up before](https://github.com/mdp-toolkit/mdp-toolkit/pull/54#issuecomment-520524879):

> I did a test using the repo [at this point in time](https://github.com/mdp-toolkit/mdp-toolkit/tree/5ffe1df89156ff84a2927da7c4304f1512f95c88). Namely, I ran `test_ccipcanode_v2` 400 times. It failed 9 times. Lets assume the repo had only this test. As we currently have 26 jobs running in Travis and given they are independent, to reach a 0.8 probability of a success of a travis run, we would need to have a probability of a single instance of this test succeeding of at least
> 
> ![](https://camo.githubusercontent.com/5b495ab857bb2a9d1d2f76c5b340932a2c0e90b1/68747470733a2f2f6c617465782e636f6465636f67732e636f6d2f6769662e6c617465783f253543736d616c6c25323071253230253341253344253230302e3825354525374225354366726163253742312537442537423236253744253744253230253344253230302e393931343534332e)
> 
> Assuming we have a probability of the test succeeding of at least that, the probability of observing 9 instances or more out of 400 failing is smaller than
> ![](https://camo.githubusercontent.com/386a6331e01f497f1500dea3ee667fe8c98d48f6/68747470733a2f2f6c617465782e636f6465636f67732e636f6d2f6769662e6c617465783f253543736d616c6c2532302535436d6174686262253742502537445f712535436269676725323825354373756d5f2537426b25334431253744253545253742343030253744253230585f69253543676571253230392535436269676725323925334425354373756d5f2537426b2533443925374425354525374234303025374425354362696e6f6d2537423430302537442537426b253744253230253238312d712532392535456b253230712535452537423430302d6b253744253230253344253230302e3030383237313233253230253543746578742e)
> 
> This suggests a more general guideline should exists for the randomized tests, especially considering that we have multiple independent randomized tests, not only this one. This fix does not live up to these standards. However, it fixes the errors for now.

**Comment by otizonaizit** [comment]
I guess that if my proposed simplification of the test environment in #74 is accepted, most of this problems will go away. It doesn't make much sense to test new Pythons with old libraries: what kind of weird setup up would it be? Also, the problems with the `numx` backend are, in my opinion, not really relevant anymore. Nowdays installing `scipy` is easy on all platforms, so there is no reason not to depend on it directly. The `numx` backend is a relic of the past that we can get rid of.

**Comment by NiMlr** [comment]
I hope a new testing setup will get rid of these problems.

Would the `numx` backend make it easier to integrate heterogeneous numerical computing capability in the future? We had thought about developing enhancements for several algorithms before.

**Comment by otizonaizit** [comment]
Oh well, the `numx` backend was introduced when `numpy` still did not exist. Back then there were two competing numerical frameworks `Numeric` and `numarray`. We wanted to support both, that is why we introduced the `numx` thing.
After `Numeric` and `numarray` were replaced by `numpy`, we started using the `numx` trick to support transparently the case where you install `numpy` but not `scipy`. Back then `scipy ` was very hard to install, so there was a clear need of only supporting numpy. Now, I think, we could get rid of the whole thing, because we can require people to install `scipy`: it is easy on all operating systems.
I don't think the `numx` thing is going to be useful for anything else.
Of course, we can remove it later, for this release it does not hurt.
